File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/iam/model/specification/FilterByCampaignIdTest.java
package com.emarsys.mobileengage.iam.model.specification;

import android.content.Context;
import android.support.test.InstrumentationRegistry;

import com.emarsys.core.database.DatabaseContract;
import com.emarsys.core.database.helper.CoreDbHelper;
import com.emarsys.core.database.helper.DbHelper;
import com.emarsys.core.database.repository.specification.QueryAll;
import com.emarsys.core.database.trigger.TriggerKey;
import com.emarsys.mobileengage.iam.model.buttonclicked.ButtonClicked;
import com.emarsys.mobileengage.iam.model.buttonclicked.ButtonClickedRepository;
import com.emarsys.mobileengage.iam.model.displayediam.DisplayedIam;
import com.emarsys.mobileengage.iam.model.displayediam.DisplayedIamRepository;
import com.emarsys.testUtil.DatabaseTestUtils;
import com.emarsys.testUtil.TimeoutUtils;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

import static org.junit.Assert.assertEquals;

public class FilterByCampaignIdTest {

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();
    private DisplayedIamRepository displayedIamRepository;
    private ButtonClickedRepository buttonClickedRepository;

    @Before
    public void init() {
        DatabaseTestUtils.deleteCoreDatabase();

        Context context = InstrumentationRegistry.getContext();
        DbHelper dbHelper = new CoreDbHelper(context, new HashMap<TriggerKey, List<Runnable>>());

        displayedIamRepository = new DisplayedIamRepository(dbHelper);
        buttonClickedRepository = new ButtonClickedRepository(dbHelper);
    }

    @Test
    public void testExecution_displayedIam_shouldDeleteIam() {
        DisplayedIam iam1 = new DisplayedIam("campaign1", 10L);
        DisplayedIam iam2 = new DisplayedIam("campaign2", 20L);
        DisplayedIam iam3 = new DisplayedIam("campaign3", 30L);

        displayedIamRepository.add(iam1);
        displayedIamRepository.add(iam2);
        displayedIamRepository.add(iam3);

        displayedIamRepository.remove(new FilterByCampaignId("campaign2"));

        List<DisplayedIam> result = displayedIamRepository.query(new QueryAll(DatabaseContract.DISPLAYED_IAM_TABLE_NAME));
        List<DisplayedIam> expected = Arrays.asList(iam1, iam3);

        assertEquals(expected, result);
    }

    @Test
    public void testExecution_displayedIam_shouldDelete_multipleIams() {
        DisplayedIam iam1 = new DisplayedIam("campaign1", 10L);
        DisplayedIam iam2 = new DisplayedIam("campaign2", 20L);
        DisplayedIam iam3 = new DisplayedIam("campaign3", 30L);
        DisplayedIam iam4 = new DisplayedIam("campaign4", 40L);

        displayedIamRepository.add(iam1);
        displayedIamRepository.add(iam2);
        displayedIamRepository.add(iam3);
        displayedIamRepository.add(iam4);

        displayedIamRepository.remove(new FilterByCampaignId("campaign1", "campaign2"));

        List<DisplayedIam> result = displayedIamRepository.query(new QueryAll(DatabaseContract.DISPLAYED_IAM_TABLE_NAME));
        List<DisplayedIam> expected = Arrays.asList(iam3, iam4);

        assertEquals(expected, result);
    }

    @Test
    public void testExecution_displayedIam_withEmptyIdArray() {
        DisplayedIam iam1 = new DisplayedIam("campaign1", 10L);
        DisplayedIam iam2 = new DisplayedIam("campaign2", 20L);
        DisplayedIam iam3 = new DisplayedIam("campaign3", 30L);
        DisplayedIam iam4 = new DisplayedIam("campaign4", 40L);

        displayedIamRepository.add(iam1);
        displayedIamRepository.add(iam2);
        displayedIamRepository.add(iam3);
        displayedIamRepository.add(iam4);

        displayedIamRepository.remove(new FilterByCampaignId());

        List<DisplayedIam> result = displayedIamRepository.query(new QueryAll(DatabaseContract.DISPLAYED_IAM_TABLE_NAME));
        List<DisplayedIam> expected = Arrays.asList(iam1, iam2, iam3, iam4);

        assertEquals(expected, result);
    }

    @Test
    public void testExecution_buttonClicked_shouldDeleteIam() {
        ButtonClicked btn1 = new ButtonClicked("campaign1", "button1", 10L);
        ButtonClicked btn2 = new ButtonClicked("campaign1", "button3", 10L);
        ButtonClicked btn3 = new ButtonClicked("campaign2", "button10", 10L);

        buttonClickedRepository.add(btn1);
        buttonClickedRepository.add(btn2);
        buttonClickedRepository.add(btn3);

        buttonClickedRepository.remove(new FilterByCampaignId("campaign2"));

        List<ButtonClicked> result = buttonClickedRepository.query(new QueryAll(DatabaseContract.BUTTON_CLICKED_TABLE_NAME));
        List<ButtonClicked> expected = Arrays.asList(btn1, btn2);

        assertEquals(expected, result);
    }

    @Test
    public void testExecution_buttonClicked_shouldDelete_multipleIams() {
        ButtonClicked btn1 = new ButtonClicked("campaign1", "button1", 10L);
        ButtonClicked btn2 = new ButtonClicked("campaign1", "button3", 10L);
        ButtonClicked btn3 = new ButtonClicked("campaign2", "button10", 10L);
        ButtonClicked btn4 = new ButtonClicked("campaign3", "button10", 10L);

        buttonClickedRepository.add(btn1);
        buttonClickedRepository.add(btn2);
        buttonClickedRepository.add(btn3);
        buttonClickedRepository.add(btn4);

        buttonClickedRepository.remove(new FilterByCampaignId("campaign1", "campaign2"));

        List<ButtonClicked> result = buttonClickedRepository.query(new QueryAll(DatabaseContract.BUTTON_CLICKED_TABLE_NAME));
        List<ButtonClicked> expected = Collections.singletonList(btn4);

        assertEquals(expected, result);
    }

    @Test
    public void testExecution_buttonClicked_withEmptyIdArray() {
        ButtonClicked btn1 = new ButtonClicked("campaign1", "button1", 10L);
        ButtonClicked btn2 = new ButtonClicked("campaign1", "button3", 10L);
        ButtonClicked btn3 = new ButtonClicked("campaign2", "button10", 10L);

        buttonClickedRepository.add(btn1);
        buttonClickedRepository.add(btn2);
        buttonClickedRepository.add(btn3);

        buttonClickedRepository.remove(new FilterByCampaignId());

        List<ButtonClicked> result = buttonClickedRepository.query(new QueryAll(DatabaseContract.BUTTON_CLICKED_TABLE_NAME));
        List<ButtonClicked> expected = Arrays.asList(btn1, btn2, btn3);

        assertEquals(expected, result);
    }

}

File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/iam/model/specification/FilterByCampaignIdTest.kt
package com.emarsys.mobileengage.iam.model.specification

import com.emarsys.core.database.helper.CoreDbHelper
import com.emarsys.core.database.repository.specification.Everything
import com.emarsys.mobileengage.iam.model.buttonclicked.ButtonClicked
import com.emarsys.mobileengage.iam.model.buttonclicked.ButtonClickedRepository
import com.emarsys.mobileengage.iam.model.displayediam.DisplayedIam
import com.emarsys.mobileengage.iam.model.displayediam.DisplayedIamRepository
import com.emarsys.testUtil.DatabaseTestUtils
import com.emarsys.testUtil.InstrumentationRegistry
import com.emarsys.testUtil.TimeoutUtils
import io.kotlintest.shouldBe
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import java.util.*

class FilterByCampaignIdTest {

    @Rule
    @JvmField
    var timeout: TestRule = TimeoutUtils.timeoutRule
    
    private lateinit var displayedIamRepository: DisplayedIamRepository
    private lateinit var buttonClickedRepository: ButtonClickedRepository

    @Before
    fun init() {
        DatabaseTestUtils.deleteCoreDatabase()

        val context = InstrumentationRegistry.getTargetContext()
        val dbHelper = CoreDbHelper(context, HashMap())

        displayedIamRepository = DisplayedIamRepository(dbHelper)
        buttonClickedRepository = ButtonClickedRepository(dbHelper)
    }

    @Test
    fun testQuery_displayedIam() {
        val iam1 = DisplayedIam("campaign1", 10L)
        val iam2 = DisplayedIam("campaign2", 20L)
        val iam3 = DisplayedIam("campaign3", 30L)

        displayedIamRepository.add(iam1)
        displayedIamRepository.add(iam2)
        displayedIamRepository.add(iam3)

        val result = displayedIamRepository.query(FilterByCampaignId("campaign2", "campaign3"))
        val expected = listOf(iam2, iam3)

        result shouldBe expected
    }

    @Test
    fun testDelete_displayedIam_shouldDeleteIam() {
        val iam1 = DisplayedIam("campaign1", 10L)
        val iam2 = DisplayedIam("campaign2", 20L)
        val iam3 = DisplayedIam("campaign3", 30L)

        displayedIamRepository.add(iam1)
        displayedIamRepository.add(iam2)
        displayedIamRepository.add(iam3)

        displayedIamRepository.remove(FilterByCampaignId("campaign2"))

        val result = displayedIamRepository.query(Everything())
        val expected = listOf(iam1, iam3)

        result shouldBe expected
    }

    @Test
    fun testDelete_displayedIam_shouldDelete_multipleIams() {
        val iam1 = DisplayedIam("campaign1", 10L)
        val iam2 = DisplayedIam("campaign2", 20L)
        val iam3 = DisplayedIam("campaign3", 30L)
        val iam4 = DisplayedIam("campaign4", 40L)

        displayedIamRepository.add(iam1)
        displayedIamRepository.add(iam2)
        displayedIamRepository.add(iam3)
        displayedIamRepository.add(iam4)

        displayedIamRepository.remove(FilterByCampaignId("campaign1", "campaign2"))

        val result = displayedIamRepository.query(Everything())
        val expected = listOf(iam3, iam4)

        result shouldBe expected
    }

    @Test
    fun testDelete_displayedIam_withEmptyIdArray() {
        val iam1 = DisplayedIam("campaign1", 10L)
        val iam2 = DisplayedIam("campaign2", 20L)
        val iam3 = DisplayedIam("campaign3", 30L)
        val iam4 = DisplayedIam("campaign4", 40L)

        displayedIamRepository.add(iam1)
        displayedIamRepository.add(iam2)
        displayedIamRepository.add(iam3)
        displayedIamRepository.add(iam4)

        displayedIamRepository.remove(FilterByCampaignId())

        val result = displayedIamRepository.query(Everything())
        val expected = listOf(iam1, iam2, iam3, iam4)

        result shouldBe expected
    }

    @Test
    fun testQuery_buttonClicked() {
        val btn1 = ButtonClicked("campaign1", "button1", 10L)
        val btn2 = ButtonClicked("campaign1", "button3", 10L)
        val btn3 = ButtonClicked("campaign2", "button10", 10L)

        buttonClickedRepository.add(btn1)
        buttonClickedRepository.add(btn2)
        buttonClickedRepository.add(btn3)

        val result = buttonClickedRepository.query(FilterByCampaignId("campaign1"))
        val expected = listOf(btn1, btn2)

        result shouldBe expected
    }

    @Test
    fun testDelete_buttonClicked_shouldDeleteIam() {
        val btn1 = ButtonClicked("campaign1", "button1", 10L)
        val btn2 = ButtonClicked("campaign1", "button3", 10L)
        val btn3 = ButtonClicked("campaign2", "button10", 10L)

        buttonClickedRepository.add(btn1)
        buttonClickedRepository.add(btn2)
        buttonClickedRepository.add(btn3)

        buttonClickedRepository.remove(FilterByCampaignId("campaign2"))

        val result = buttonClickedRepository.query(Everything())
        val expected = listOf(btn1, btn2)

        result shouldBe expected
    }

    @Test
    fun testDelete_buttonClicked_shouldDelete_multipleIams() {
        val btn1 = ButtonClicked("campaign1", "button1", 10L)
        val btn2 = ButtonClicked("campaign1", "button3", 10L)
        val btn3 = ButtonClicked("campaign2", "button10", 10L)
        val btn4 = ButtonClicked("campaign3", "button10", 10L)

        buttonClickedRepository.add(btn1)
        buttonClickedRepository.add(btn2)
        buttonClickedRepository.add(btn3)
        buttonClickedRepository.add(btn4)

        buttonClickedRepository.remove(FilterByCampaignId("campaign1", "campaign2"))

        val result = buttonClickedRepository.query(Everything())
        val expected = listOf(btn4)

        result shouldBe expected
    }

    @Test
    fun testDelete_buttonClicked_withEmptyIdArray() {
        val btn1 = ButtonClicked("campaign1", "button1", 10L)
        val btn2 = ButtonClicked("campaign1", "button3", 10L)
        val btn3 = ButtonClicked("campaign2", "button10", 10L)

        buttonClickedRepository.add(btn1)
        buttonClickedRepository.add(btn2)
        buttonClickedRepository.add(btn3)

        buttonClickedRepository.remove(FilterByCampaignId())

        val result = buttonClickedRepository.query(Everything())
        val expected = listOf(btn1, btn2, btn3)

        result shouldBe expected
    }

}